### DISCUSSION TITLE
Liquidsoap 2.2.1 blank.strip with mksafe combo causes CPU increase triggering 'we must catchup' events

### DISCUSSION THREAD

**Discussion by SkillenUK** [savonet/liquidsoap#3474] [User Support]
I use a fallback between two input harbors, one for ssl and one for non ssl clients. I then use mksafe so that the listener initially hears a blank when the DJ disconnects. I use blank.strip to then move to the playlist after 10 seconds of blank. I have used this in my script for many years, but since 2.2.x, the CPU starts spiking whilst playing the playlist (autodj) until it causes the stream to break up with 'We must catchup x seconds!' events appearing in the log.

I use this in my script so that if the DJ has a bad connection and drops off for a few seconds while they are streaming then the auto DJ (playlist) doesn't instantly come in. It will also put the playlist on after 10 seconds if the DJ connects but is streaming a blank output.

```
cert = http.transport.ssl(
  certificate="/etc/ssl/certs/cert.crt",
  key="/etc/ssl/private/key.key")

def dj_auth(params)
  user     = params.user
  password = params.password

  ret = process.read.lines("php /path/to/liqauth.php '#{user}' '#{password}'")
  ret = list.hd(default="",ret)

  if ret == "true" then
    true
  else
    false
  end
end

djlivessl = input.harbor(transport=cert,"live",port=8020,auth=dj_auth,buffer=2.)
djlive = input.harbor("live",port=8010,password="hackme",buffer=2.)

live = blank.strip(max_blank=10.,mksafe(fallback(track_sensitive=false, [djlivessl,djlive])))
autodj =  playlist(mode="randomize",reload=1,reload_mode="rounds","autodj.m3u")

mainoutput =  mksafe(fallback(track_sensitive=false, transition_length=10.,transitions=[jinglenew], [(live:source),(autodj:source)]))

output.icecast(
  %mp3(bitrate=192),
  mount="/stream192",
  host="localhost", port=8000, password="hackme",
  name="Stream", description="Main 192kbps MP3 stream", url="https://www.url.net", genre="Genre", mainoutput)
```

Basically after replacing `live = blank.strip(max_blank=10.,mksafe(fallback(track_sensitive=false, [djlivessl,djlive])))` with `live = fallback(track_sensitive=false, [djlivessl,djlive])` in my script the problem goes away, so I think this is the key. Strangely it is while the playlist is playing (not the input harbour) that the problem manifests and it happens within a few minutes.

I suspect there's a different way of achieving the same thing, so if you can suggest a work-around I would be grateful.

I'm running Liquidsoap 2.2.1 in Ubuntu 22.04.

Apologies for editing this a few times, I accidentally posted it before typing the conent out :)

Thanks, Skillen.

**Comment by toots** [comment]
I don't know why it was working before but you should not use `blank.strip` with `mksafe`.

The reason is that:
* `blank.strip` is designed to _consume all available data_ until some non-silent audio is detected. This is also an active source so it is always animated even when not connected to an output. This is required to make sure it becomes available again the next time the underlying source does not have silent audio.
* `mksafe` is designed to replace unavailable sources with _slient audio_

The combination of the two is pretty dangerous as you end up in a situation when, if the `input.harbor` is not available, `mksafe` returns silent audio and `blank.strip` keeps consuming it, leading to a CPU spike.

Thus, you should remove the `mksafe` under `blank.strip` and everything should work as expected.

**Comment by yan-sag** [comment]
Hi @toots ,
It seems I have a similar issue with `blank.strip`.

I used to use `strip_blank(mksafe(s))` with 1.x versions. Since I use 2.2.2, I rewrited my scripts and deleted `blank.strip(mksafe(s))`.
I still have an increasing %CPU, from ~6% to 100+% leading to 'must catch-up' errors as soon as I use the `blank.strip` operator.

I tested with the minimal code, you gave in The Liquidsoap book, the CPU load increases when it switch to `music` source, and decreases when it switch to `mic`.

```
mic = buffer(input.alsa(device="plughw:1,0"))
mic = blank.strip(max_blank=2., min_noise=.1, threshold=-20., mic)
music = playlist("~/musique")
radio = fallback(track_sensitive=false, [mic, music])

output.alsa(fallible=true, device='plughw:1,0', radio)
```

**Reply by yan-sag** [reply]
Just tested with 2.1.4, average %CPU on `music` or `mic` : 3.5%

**Comment by AVK-RTX** [comment]
Hi,
I have a similar issue with versions 2.2.1 and 2.2.2.
With blank.strip CPU usage quickly increases from 10% to 100%.
On 2.1.4 version blank.strip work greate.

my test config:

line_in = buffer(input.alsa(device="plughw:0,0"))
line_in = blank.strip(max_blank=10., min_noise=10., threshold=-50., line_in)
air_rezerv = sine(id="sine_test", amplitude=0.8, 1000.)
air_out = fallback(track_sensitive=false, [line_in, air_rezerv])

output.alsa(fallible=true, bufferize=false, device="plughw:0,0",  air_out)

**Comment by toots** [comment]
This should be fixed in `rolling-release-v2.2.x`: https://github.com/savonet/liquidsoap/releases/tag/rolling-release-v2.2.x we'll do a stable release with very soon.

**Reply by AVK-RTX** [reply]
Thanks for the bug fix! I tested for ~48 hours on version 2.2.3 (2023-12-14) . There are no problems with increasing cpu. Good job!

**Comment by SkillenUK** [comment]
I've tested in 2.2.3 as well and it appears to be fine now, it will be great to have that resolved for when I can move on to the latest version.

Unfortunately I can't use 2.2.x at the moment because the [metadata issue](https://github.com/savonet/liquidsoap/issues/3574) which I reported, it's quite fundamental to the way which we use Liquidsoap. But thanks for fixing this all the same :)

Skillen.
